crates/datasources/src/bson/builder.rs

```diff
@@ -90,27 +90,26 @@ impl RecordStructBuilder {
 
     pub fn append_record(&mut self, doc: &RawDocument) -> Result<()> {
         let mut cols_set: BitVec<u8, Lsb0> = BitVec::repeat(false, self.fields.len());
+        let doc_buf = doc.to_raw_document_buf();
```

Comment:
This seems to copy the underlying bytes.

Are we sure we're actually improving an average performance by implementing this (compared to the extra alloc and copy of bytes)? This seems like a guess.

---

I did look into the underlying iterator. `RawIter::new` is not public. We can raise a PR to make it public since it accepts `&RawDocument` and this could be avoided. Until then, I don't think we should move ahead with this without any concrete benchmarks.

Reply:
I pushed a patch that avoids the extra copy (except for nested documents, but will open a couple of ergonomic features on the upstream lib.)